vm/src/types/program.rs

```diff
@@ -98,6 +95,78 @@ impl<'a> Arbitrary<'a> for SharedProgramData {
     }
 }
 
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub(crate) struct HintsCollection {
+    hints: Vec<HintParams>,
+    /// This maps a PC to the range of hints in `hints` that correspond to it.
+    hints_ranges: Vec<HintRange>,
+}
+
+impl HintsCollection {
+    pub(crate) fn new(
+        hints: HashMap<usize, Vec<HintParams>>,
+        program_length: usize,
+    ) -> Result<Self, ProgramError> {
+        let bounds = hints
+            .iter()
+            .map(|(pc, hs)| (*pc, hs.len()))
+            .reduce(|(max_hint_pc, full_len), (pc, len)| (max_hint_pc.max(pc), full_len + len));
+
+        let Some((max_hint_pc, full_len)) = bounds else {
+            return Ok(HintsCollection {
+                hints: Vec::new(),
+                hints_ranges: Vec::new(),
+            });
+        };
+
+        if max_hint_pc >= program_length {
+            return Err(ProgramError::InvalidHintPc(max_hint_pc, program_length));
+        }
+
+        let mut hints_values = Vec::with_capacity(full_len);
+        let mut hints_ranges = vec![None; max_hint_pc + 1];
+
+        for (pc, hs) in hints.iter().filter(|(_, hs)| !hs.is_empty()) {
+            let range = (
+                hints_values.len(),
+                NonZeroUsize::new(hs.len()).expect("empty vecs already filtered"),
+            );
+            hints_ranges[*pc] = Some(range);
+            hints_values.extend_from_slice(&hs[..]);
+        }
+
+        Ok(HintsCollection {
+            hints: hints_values,
+            hints_ranges,
+        })
+    }
+
+    pub fn iter_hints(&self) -> impl Iterator<Item = &HintParams> {
+        self.hints.iter()
+    }
+
+    pub fn get_hint_range_for_pc(&self, pc: usize) -> Option<HintRange> {
+        self.hints_ranges.get(pc).cloned()
+    }
```

Comment:
What I was thinking of was something more like:
```suggestion
    pub fn get_hint_range_for_pc(&self, pc: usize) -> Option<&[HintParams]> {
        let bounds = self.hints_ranges.get(pc)?;
        self.hints.get(bounds.0..bounds.0+bounds.1)
    }
```
This is directly usable later on.